curl_cffi/requests/websockets.py

```diff
@@ -33,30 +64,269 @@ class WsCloseCode(IntEnum):
 
 
 class WebSocketError(CurlError):
-    pass
+    """WebSocket-specific error."""
+
+    def __init__(self, message: str, code: Union[WsCloseCode, CurlECode, Literal[0]] = 0):
+        super().__init__(message, code)  # type: ignore
+
+
+async def aselect(fd, *, loop: asyncio.AbstractEventLoop, timeout: Optional[float] = None) -> bool:
+    future = loop.create_future()
+    loop.add_reader(fd, future.set_result, None)
+    future.add_done_callback(lambda _: loop.remove_reader(fd))
+    try:
+        await asyncio.wait_for(future, timeout)
+    except asyncio.TimeoutError:
+        return False
+    return True
+
+
+class BaseWebSocket:
+    def __init__(self, curl: Curl, *, autoclose: bool = True):
+        self.curl: Curl = curl
+        self.autoclose: bool = autoclose
+        self._close_code: Optional[int] = None
+        self._close_reason: Optional[str] = None
+
+    @property
+    def closed(self) -> bool:
+        """Whether the WebSocket is closed."""
+        return self.curl is not_set
+
+    @property
+    def close_code(self) -> Optional[int]:
+        """The WebSocket close code, if the connection is closed."""
+        return self._close_code
+
+    @property
+    def close_reason(self) -> Optional[str]:
+        """The WebSocket close reason, if the connection is closed."""
+        return self._close_reason
+
+    @staticmethod
+    def _pack_close_frame(code: int, reason: bytes) -> bytes:
+        return struct.pack("!H", code) + reason
+
+    @staticmethod
+    def _unpack_close_frame(frame: bytes) -> Tuple[int, str]:
+        if len(frame) < 2:
+            code = WsCloseCode.UNKNOWN
+            reason = ""
+        else:
+            try:
+                code = struct.unpack_from("!H", frame)[0]
+                reason = frame[2:].decode()
+            except UnicodeDecodeError:
+                raise WebSocketError("Invalid close message", WsCloseCode.INVALID_DATA)
+            except Exception:
+                raise WebSocketError("Invalid close frame", WsCloseCode.PROTOCOL_ERROR)
+            else:
+                if code < 3000 and (code not in WsCloseCode or code == 1005):
+                    raise WebSocketError("Invalid close code", WsCloseCode.PROTOCOL_ERROR)
+        return code, reason
+
+    def terminate(self):
+        """Terminate the underlying connection."""
+        if self.curl is not_set:
+            return
+        self.curl.close()
+        self.curl = not_set
 
 
-class WebSocket:
+class WebSocket(BaseWebSocket):
+    """A WebSocket implementation using libcurl."""
+
     def __init__(
         self,
-        session,
-        curl,
-        on_message: Optional[ON_MESSAGE_T] = None,
-        on_error: Optional[ON_ERROR_T] = None,
+        *,
```

Comment:
It's possible but would make the overall API worse imo, because currently the `WebSocket` class fulfills the role of both `WebSocketApp` and `create_connection()`, and the user can choose to operate it by either calling `connect()` and manually receiving or calling `run_forever()` and having a continuous loop. Putting all the parameters in the init would make the mechanics of `connect()` kinda weird.